src/main/java/com/axonivy/market/service/impl/ProductServiceImpl.java

```diff
@@ -0,0 +1,251 @@
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.domain.Sort.Order;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+import com.axonivy.market.constants.GitHubConstants;
+import com.axonivy.market.entity.GithubRepoMeta;
+import com.axonivy.market.entity.Product;
+import com.axonivy.market.enums.FileType;
+import com.axonivy.market.enums.FilterType;
+import com.axonivy.market.enums.SortOption;
+import com.axonivy.market.factory.ProductFactory;
+import com.axonivy.market.github.model.GitHubFile;
+import com.axonivy.market.github.service.GHAxonIvyMarketRepoService;
+import com.axonivy.market.github.service.GithubService;
+import com.axonivy.market.github.util.GithubUtils;
+import com.axonivy.market.repository.GithubRepoMetaRepository;
+import com.axonivy.market.repository.ProductRepository;
+import com.axonivy.market.service.ProductService;
+
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Service
+public class ProductServiceImpl implements ProductService {
+
+  private final ProductRepository productRepository;
+  private final GHAxonIvyMarketRepoService axonivyMarketRepoService;
+  private final GithubRepoMetaRepository githubRepoMetaRepository;
+  private final GithubService githubService;
+
+  private GHCommit lastGHCommit;
+  private GithubRepoMeta marketRepoMeta;
+
+  public ProductServiceImpl(ProductRepository productRepository, GHAxonIvyMarketRepoService axonivyMarketRepoService,
+      GithubRepoMetaRepository githubRepoMetaRepository, GithubService githubService) {
+    this.productRepository = productRepository;
+    this.axonivyMarketRepoService = axonivyMarketRepoService;
+    this.githubRepoMetaRepository = githubRepoMetaRepository;
+    this.githubService = githubService;
+  }
+
+  @Override
+  public Page<Product> findProducts(String type, String keyword, Pageable pageable) {
+    final var filterType = FilterType.of(type);
+    if (StringUtils.isNoneBlank(keyword)) {
+      return searchProducts(filterType, keyword, pageable);
+    }
+
+    syncLatestDataFromMarketRepo();
+
+    Pageable unifiedPageabe = refinePagination(pageable);
+    return switch (filterType) {
+    case ALL -> productRepository.findAll(unifiedPageabe);
+    case CONNECTORS, UTILITIES, SOLUTIONS -> productRepository.findByType(filterType.getCode(), pageable);
+    default -> Page.empty();
+    };
+  }
+
+  @Override
+  public boolean syncLatestDataFromMarketRepo() {
+    var isAlreadyUpToDate = isLastGithubCommitCovered();
+    if (!isAlreadyUpToDate) {
+      if (marketRepoMeta == null) {
+        syncProductsFromGithubRepo();
+        marketRepoMeta = new GithubRepoMeta();
+      } else {
+        updateLatestChangeToProductsFromGithubRepo();
+      }
+      syncRepoMetaDataStatus();
+    }
+    return isAlreadyUpToDate;
+  }
+
+  private void syncRepoMetaDataStatus() {
+    if (marketRepoMeta == null || lastGHCommit == null) {
```

Comment:
MarketRepoMeta is always not null. Do we need the condition marketRepoMeta == null?